File: csharpfunctions/DataProcessing/DataProcessing_4.cs
using Microsoft.Azure.WebJobs;
using Microsoft.Extensions.Logging;
using Octokit;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

namespace csharpfunctions.DataProcessing
{
    public static class DataProcessing_4
    {
        //http://localhost:7071/admin/functions/DataProcessing_4
        [FunctionName("DataProcessing_4")]
        public static async Task Run([TimerTrigger("0 0 1 */1 * *")]TimerInfo myTimer,
            [CosmosDB(
                databaseName: "GithubStats",
                collectionName: "DocsRepo",
                ConnectionStringSetting = "CosmosDBConnection")]  IAsyncCollector<IssueDataV2> outputData,
            ILogger log)
        {
            log.LogInformation($"C# DataProcessing_4 function executed at: {DateTime.Now}");

            var client = new GitHubClient(new ProductHeaderValue("fsharpfunctions"));
            var tokenAuth = new Credentials(Environment.GetEnvironmentVariable("GithubToken")); // NOTE: not real token
            client.Credentials = tokenAuth;

            log.LogInformation("Retrieving Github Issues");

            var request = new SearchIssuesRequest()
            {
                PerPage = 100,
                Page = 1,
                Type = IssueTypeQualifier.Issue,
                State = ItemState.Open
            };
            request.Repos.Add("MicrosoftDocs", "azure-docs");
            var openIssues = await client.Search.SearchIssues(request);
            log.LogInformation($"Retrieved {openIssues.Items.Count()} Issues"); ;

            log.LogInformation(client.GetLastApiInfo().RateLimit.Remaining.ToString());

            var byPriority = openIssues.Items
                .SelectMany(o => o.Labels)
                .Where(l => l.Name.Contains("Pri"))
                .GroupBy(p => p.Name)
                .ToDictionary(g => g.Key, g => g.Count());

            var byService = openIssues.Items
                .SelectMany(o => o.Labels)
                .Where(l => l.Name.EndsWith("/svc"))
                .GroupBy(p => p.Name.CleanServiceName())
                .ToDictionary(g => g.Key, g => g.Count());

            var serviceCount = new Dictionary<string, ItemCount>();

            foreach(var item in openIssues.Items)
            {
                var serviceLabel = item.Labels.Where(l => l.Name.EndsWith("/svc")).FirstOrDefault();
                if (serviceLabel != null)
                {
                    var serviceName = serviceLabel.Name.CleanServiceName();
                    if (!serviceCount.TryAdd(serviceName, new ItemCount() { Name = serviceName, Count = 1 }))
                    {
                        serviceCount[serviceName].Count++;
                    }

                    var subserviceLabel = item.Labels.Where(l => l.Name.EndsWith("/subsvc")).FirstOrDefault();
                    if (subserviceLabel != null)
                    {
                        var subserviceName = subserviceLabel.Name.CleanServiceName();

                        if (serviceCount[serviceName].Children == null)
                        {
                            serviceCount[serviceName].Children = new List<ItemCount>();
                        }

                        if (serviceCount[serviceName].Children.Where(s => s.Name == subserviceName).Any())
                        {

                            var index = serviceCount[serviceName].Children.IndexOf(new ItemCount() { Name = subserviceName });
                            if(index != -1)
                            {
                                serviceCount[serviceName].Children[index].Count++;
                            }
                        }
                        else
                        {
                            serviceCount[serviceName].Children.Add(new ItemCount() { Name = subserviceName, Count = 1 });
                        }
                    }
                }
            }

            var document = new IssueDataV2()
            {
                Id = Guid.NewGuid().ToString(),
                Source = "CSharp",
                EntryType = "BySubService",
                Timestamp = DateTime.UtcNow,
                TotalOpenIssues = openIssues.Items.Count(),
                MissingTags = openIssues.Items.Count(o => o.Labels.Count() == 0),
                CountByPriority = byPriority,
                CountByService = serviceCount.Values.ToList()
            };

            await outputData.AddAsync(document);
        }

        private static string CleanServiceName(this string str)
        {
            return str.Substring(0, str.LastIndexOf('/'));
        }
    }
}

File: fsharpfunctions/DataProcessing/DataProcessing_4.fs
namespace FSharpFunctions

open System
open Microsoft.Azure.WebJobs
open Microsoft.Extensions.Logging
open Octokit
open Microsoft.FSharpLu.Json

module DataProcessing_4 =
    let createClient headerValue token = 
        let client = ProductHeaderValue headerValue |> GitHubClient
        client.Credentials <- Credentials token
        client 
    
    let tagFilter (issue: Issue, searchString: string) =
        (issue.Labels) 
            |> Seq.find(fun f -> f.Name.Contains(searchString))
    let getFirstTag(issue: Issue, searchString: string) =
        (issue.Labels) 
            |> Seq.tryPick(fun f -> 
                if f.Name.Contains(searchString) then
                    Some(f)
                else None)
    let cleanName (tag: Label option) =
        match tag with 
         | Some t -> Some (t.Name.Substring(0, t.Name.LastIndexOf('/')))
         | None -> None
     
    let subServiceSeqToOptionList (os: seq<SubService option>): list<SubService> option = 
        os |> Seq.choose id |> Seq.toList |> Some


    [<FunctionName("DataProcessing_4")>]
    let run([<TimerTrigger("0 0 0 */1 * *")>]myTimer: TimerInfo, [<CosmosDB(databaseName= "GithubStats", collectionName= "DocsRepo", ConnectionStringSetting = "CosmosDBConnection")>]  outputData: IAsyncCollector<string>,  log: ILogger) =
        async {
            sprintf "F# DataProcessing_4 function executed at: %A" DateTime.Now |> log.LogInformation 

            let token = Environment.GetEnvironmentVariable "GithubToken"
            let client = createClient "fsharpfunctions" token

            log.LogInformation("Retrieving Github Issues")
            let request = SearchIssuesRequest( PerPage = 100, Page = 1, Type = Nullable IssueTypeQualifier.Issue, State =  Nullable ItemState.Open)
            request.Repos.Add("MicrosoftDocs", "azure-docs");
            let! openIssues = client.Search.SearchIssues request |> Async.AwaitTask

            sprintf "Retrieved %A Issues" openIssues.Items.Count |> log.LogInformation
            client.GetLastApiInfo().RateLimit.Remaining.ToString() |> log.LogInformation

            let issues = (openIssues.Items) |> Seq.cache

            let noTags = issues |> Seq.filter (fun f-> f.Labels.Count = 0)
            let byPriority = issues |> Seq.countBy (fun f -> 
                let tag = tagFilter (f, "Pri") 
                tag.Name )

            let byService = 
                issues 
                |> Seq.map (fun f ->
                    let service = getFirstTag (f, "/svc")
                    match service with
                        | Some x -> (x.Name, getFirstTag (f, "/subsvc") |> cleanName)
                        | None -> ("Unknown", getFirstTag (f, "/subsvc") |> cleanName)
                )
                |> Seq.groupBy (fst)
                |> Seq.map ( fun f -> {
                    Name = if (fst f).LastIndexOf('/') > 0 then (fst f).Substring(0, (fst f).LastIndexOf('/')) else fst f;
                    Count = snd f |> Seq.length;
                    SubServiceCounts = 
                        snd f
                        |> Seq.countBy ( fun g -> snd g)
                        |> Seq.map (fun h -> 
                            match fst h with
                                | Some x -> {Name = x; Count = snd h;} |> Some
                                | None -> None)
                        |> subServiceSeqToOptionList
                }) |> Seq.toList |> Some
          

            let document = {
                IssueDataV2.Id = Guid.NewGuid().ToString(); 
                Source = FSharp; 
                EntryType = ByService; 
                Timestamp = DateTime.UtcNow;
                TotalOpenIssues = openIssues.Items.Count;
                MissingTags = Seq.length noTags;
                CountByPriority = Some ( byPriority  |> Map.ofSeq );
                CountByService = byService;
            }

            do! Compact.serialize document |> outputData.AddAsync |> Async.AwaitTask
        } |> Async.StartAsTask

